diff --git a/mrbgems/picoruby-filesystem-fat/ports/rp2040/flash_disk.c b/mrbgems/picoruby-filesystem-fat/ports/rp2040/flash_disk.c
index 78fde305..07792c35 100644
--- a/mrbgems/picoruby-filesystem-fat/ports/rp2040/flash_disk.c
+++ b/mrbgems/picoruby-filesystem-fat/ports/rp2040/flash_disk.c
@@ -9,6 +9,25 @@
 
 #include "disk.h"
 
+static int is_rp2040js() {
+  // Read SYSINFO PLATFORM register
+  const uint32_t* p = (const uint32_t*)(0x40000000 + 0x4);
+  return (*p & 0x01000000) != 0;
+}
+
+static void break_flash_range_program(uint32_t flash_offs, const uint8_t *data, size_t count) {
+  asm("bkpt 27");
+  return;
+}
+
+static void custom_flash_range_program(uint32_t flash_offs, const uint8_t *data, size_t count) {
+  if (is_rp2040js()) {
+    break_flash_range_program(flash_offs, data, count);
+  } else {
+    flash_range_program(flash_offs, data, count);
+  }
+}
+
 int
 FLASH_disk_erase(void)
 {
@@ -64,7 +83,7 @@ FLASH_disk_write(const BYTE *buff, LBA_t sector, UINT count)
   /*
    * Program (Write) unit should be a multiple of FLASH_PAGE_SIZE (256 Byte)
    */
-  flash_range_program(
+  custom_flash_range_program(
     (uint32_t)(FLASH_TARGET_OFFSET + sector * FLASH_SECTOR_SIZE),
     (const uint8_t *)buff,
     (size_t)(FLASH_SECTOR_SIZE * count)
