diff --git a/mrbgems/picoruby-psg/ports/rp2040/psg_port.c b/mrbgems/picoruby-psg/ports/rp2040/psg_port.c
index 674930b6..958815a2 100644
--- a/mrbgems/picoruby-psg/ports/rp2040/psg_port.c
+++ b/mrbgems/picoruby-psg/ports/rp2040/psg_port.c
@@ -38,16 +38,18 @@ psg_port_init(void)
 psg_cs_token_t
 PSG_enter_critical(void)
 {
-  uint32_t irq_state = save_and_disable_interrupts();
-  spin_lock_unsafe_blocking(psg_spin);
-  return irq_state; // token = IRQ
+  return spin_lock_blocking(psg_spin);
+  // uint32_t irq_state = save_and_disable_interrupts();
+  // spin_lock_unsafe_blocking(psg_spin);
+  // return irq_state; // token = IRQ
 }
 
 void
 PSG_exit_critical(psg_cs_token_t token)
 {
-  spin_unlock(psg_spin);
-  restore_interrupts(token);
+  spin_unlock(psg_spin, token);
+  // spin_unlock_unsafe(psg_spin);
+  // restore_interrupts(token);
 }
 
 #elif defined(PICO_RP2350)
@@ -127,7 +129,7 @@ psg_core1_main(void)
   if (!tick_alarm_pool) {
     tick_alarm_pool = alarm_pool_create(ALARM_TICK, 2);
     assert(tick_alarm_pool && "Failed to create alarm tick_alarm_pool");
-    irq_set_priority(TIMER1_IRQ_2, 3);
+    irq_set_priority(TIMER_IRQ_2, 3);
   }
   memset(&tick_timer, 0, sizeof(repeating_timer_t));
   /* 1 kHz = -1000 µs */
@@ -217,7 +219,7 @@ PSG_tick_start_core1(uint8_t p1, uint8_t p2)
   if (!audio_alarm_pool) {
     audio_alarm_pool = alarm_pool_create(ALARM_AUDIO, 2);
     assert(audio_alarm_pool && "Failed to create alarm audio_alarm_pool");
-    irq_set_priority(TIMER1_IRQ_1, 0);
+    irq_set_priority(TIMER_IRQ_1, 0);
   }
   memset(&audio_timer, 0, sizeof(repeating_timer_t));
   /* 22.05 kHz */
diff --git a/mrbgems/picoruby-shell/shell_executables/r2p2.rb b/mrbgems/picoruby-shell/shell_executables/r2p2.rb
index 66cad0c8..bf229ebf 100644
--- a/mrbgems/picoruby-shell/shell_executables/r2p2.rb
+++ b/mrbgems/picoruby-shell/shell_executables/r2p2.rb
@@ -12,5 +12,5 @@ elsif File.exist?("#{ENV['HOME']}/app.rb")
   puts "Loading app.rb"
   load "#{ENV['HOME']}/app.rb"
 else
-  puts "No app.(mrb|rb) found"
+  puts "No app.(mrb|rb) found\n"
 end
