diff --git a/demo/micropython-run.ts b/demo/micropython-run.ts
index 2eed868..bfdddaa 100644
--- a/demo/micropython-run.ts
+++ b/demo/micropython-run.ts
@@ -90,5 +90,29 @@ process.stdin.on('data', (chunk) => {
   }
 });
 
+simulator.rp2040.onBreak = (code) => {
+  // TODO: raise HardFault exception
+  // console.error('Breakpoint!', code);
+  // console.log(code);
+
+  if(code == 27){
+    const this_ = simulator.rp2040.core;
+    const RAM_START_ADDRESS = 0x20000000;
+    const flashAddr = this_.registers[0];
+    const ramAddr = this_.registers[1] - RAM_START_ADDRESS;
+    const count = this_.registers[2];
+
+    // console.log("flash write:", flashAddr, ramAddr, count)
+
+    simulator.rp2040.flash.set(simulator.rp2040.sram.slice(ramAddr, ramAddr+count), flashAddr);
+
+    // Copy LR to PC register
+    // this.registers[15] = this.registers[14];
+    // this_.PC = this_.LR;
+  } else {
+    simulator.stop();
+  }
+};
+
 simulator.rp2040.core.PC = 0x10000000;
 simulator.execute();
